libtock/crt0.c

```diff
@@ -63,8 +63,19 @@ void _start(void* app_start,
   // bss_size from start of memory. Also make sure that the stack starts on an
   // 8 byte boundary per section 5.2.1.2 here:
   // http://infocenter.arm.com/help/topic/com.arm.doc.ihi0042f/IHI0042F_aapcs.pdf
-  struct hdr* myhdr = (struct hdr*)app_start;
-  uint32_t stacktop = (((uint32_t)mem_start + myhdr->stack_size + 7) & 0xfffffff8);
+
+  // Before adjusting the stack, make sure to keep all variables in registers.
+  register struct hdr* myhdr = (struct hdr*)app_start;
+  register uint32_t stacktop = (((uint32_t)mem_start + myhdr->stack_size + 7) & 0xfffffff8);
```

Comment:
I'm pretty sure the register keyword is just a hint (and inhibits taking the variable's memory address). I think if we _actually_ want to force this to live in registers only, we have to write assembly here

Reply:
so should we just do an entire `_start0` routine in assembly that allocates the stack+heap and moves the stack, and then just to a `_start1` in C that loads the data section into memory and rewrites the GOT?

Seems safest given your point, right?

Reply:
Probably. I have some flights tomorrow afternoon/evening. I can take a stab at implementing that then.